### PR TITLE
Improve generate_test_keys.py

### LINKED ISSUE #9010: Improve `generate_test_keys.py`
This is a follow up of #8986 and it's purpose is to implement suggestions provided in [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8986#pullrequestreview-1979455049):

> - Currently it's handled like tests/src/test_certs.h: the file is commited and there is no check that it's up-to-date. That doesn't look great - either the file should be auto-generated on the fly when building tests, or we should have a check that it's up-to-date. I'm OK addressing that in a follow-up since it's arguably pre-existing.
> - Currently, every time we re-generate the file (for example to add a new key) all the existing keys change. I'm thinking perhaps we should have files for each of the keys (like you had in the first commit originally), and then the script would just read those files and create macros out of their content (which would be more similar to what test_certs.h is doing). That way we could run the script twice in a row and get the same results. Again, I'm happy to leave that to a follow-up as the PR is already pretty big, and an improvement over the existing as it stands.

To automatically check the generated files we can use `check-generated-files.sh` as suggested [here](https://github.com/Mbed-TLS/mbedtls/pull/8986#issuecomment-2037910316).

### Justification

Improve management of test keys used in `test_suite_pk`.

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Resolves #9010 
~~Depends on #8986~~ (merged).

## PR checklist

- [x] **changelog** not required, only improving tests instrumentation
- [x] **3.6 backport** https://github.com/Mbed-TLS/mbedtls/pull/9102
- [x] **2.28 backport** not required
- [x] **tests** not required

### CODE REVIEW

File: tests/Makefile [hunk 1]
```diff
@@ -112,13 +112,80 @@ all: $(BINARIES)
...
+				--string TEST_CLI_CRT_EC_PEM=data_files/cli2.crt \
+				--binary TEST_CLI_CRT_EC_DER=data_files/cli2.crt.der \
+				--string TEST_CLI_KEY_EC_PEM=data_files/cli2.key \
+				--binary TEST_CLI_KEY_EC_DER=data_files/cli2.key.der \
+				--string TEST_CLI_CRT_RSA_PEM=data_files/cli-rsa-sha256.crt \
+				--binary TEST_CLI_CRT_RSA_DER=data_files/cli-rsa-sha256.crt.der \
+				--string TEST_CLI_KEY_RSA_PEM=data_files/cli-rsa.key \
+				--binary TEST_CLI_KEY_RSA_DER=data_files/cli-rsa.key.der
+
+src/test_keys.h: scripts/generate_test_keys.py
```
Comment (gilles-peskine-arm) [inline tests/Makefile:178]: This is a new generated files, so you need to register it in a few places:

* Make sure it's generated by `make generated_files`.
* A line in `tests/scripts/check-generated-files.sh` (the absence should break `all.sh check_generated_files` but it currently doesn't, I have some unfinished work to resolve that).
* `scripts/make_generated_files.bat`, which is what breaks the Windows jobs.
*

File: tests/scripts/check-python-files.sh [hunk 2]
```diff
@@ -62,6 +62,7 @@ $PYTHON -m pylint scripts/mbedtls_dev/*.py scripts/*.py tests/scripts/*.py || {
 
 echo
 echo 'Running mypy ...'
+export MYPYPATH="../../scripts"
```
Comment (gilles-peskine-arm) [inline tests/scripts/check-python-files.sh:65]: I'd prefer not to need that, and we didn't need it before. Why do we need it now?

In any case, this looks wrong: for `scripts/foo.py`, this tells mypy to look outside the mbedtls tree?

Comment (valeriosetti) [reply tests/scripts/check-python-files.sh:65]: > Why do we need it now?

mypy was complaining for not finding `asymmetric_key_data`. I think this is because `generate_test_keys.py` is in `[root]/tests/scripts`, while `asymmetric_key_data` is in `[root]/scripts/mbedtls_dev`.
Even the import on top of the file looks quite ugly:
```
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__)) + "/"
sys.path.append(SCRIPT_DIR + "../../scripts/")
from mbedtls_dev.asymmetric_key_data import ASYMMETRIC_KEY_DATA
```

but I didn't find any better solution. The only option I had in mind was to move `generate_test_keys` to `[root]/scripts/mbedtls_dev`, but that didn't look as the right place for a test script. 
Any idea is absolutely welcome :)

Comment (valeriosetti) [reply tests/scripts/check-python-files.sh]: I applied the [solution](https://github.com/Mbed-TLS/mbedtls/pull/8986#discussion_r1561519572) you proposed in #8986. Thanks again!

File: tests/Makefile [hunk 3]
```diff
@@ -112,13 +112,80 @@ all: $(BINARIES)
...
+	scripts/generate_test_keys.py --output $@
```
Comment (gilles-peskine-arm) [inline tests/Makefile:179]:
```suggestion
	$(PYTHON) scripts/generate_test_keys.py --output $@
```
for the sake of Windows, and also for users who want to override the choice of Python.

File: CMakeLists.txt [hunk 4]
```diff
@@ -311,10 +311,125 @@ add_subdirectory(pkgconfig)
...
+            ${CMAKE_CURRENT_SOURCE_DIR}/tests/scripts/generate_test_keys.py
+    )
+    add_custom_command(
+        OUTPUT
+            ./tests/src/test_certs.h
+        WORKING_DIRECTORY
+            ${CMAKE_CURRENT_SOURCE_DIR}
+        COMMAND
+            "${MBEDTLS_PYTHON_EXECUTABLE}"
+                "${CMAKE_CURRENT_SOURCE_DIR}/tests/scripts/generate_test_cert_macros.py"
```
Comment (gilles-peskine-arm) [inline CMakeLists.txt:334]: That's a very long list of parameters that needs to be repeated in many places, so I'm requesting a design change:

* `generate_test_cert_macros.py` with no arguments just works. All the file names involved and other data are stored in the script.
* `generate_test_cert_macros.py --list-dependencies` (option name TBD, I don't think we have an established standard here) lists the dependencies of the script. Call this in `Makefile` and `CMakeLists.txt`.

File: CMakeLists.txt [hunk 5]
```diff
@@ -311,10 +311,125 @@ add_subdirectory(pkgconfig)
 # to define the test executables.
 #
 if(ENABLE_TESTING OR ENABLE_PROGRAMS)
+    add_custom_command(
```
Comment (gilles-peskine-arm) [inline CMakeLists.txt:314]: All this stuff should be in `tests/CMakeLists.txt` rather than the toplevel file.

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -74,6 +74,34 @@ foreach(file ${base_psa_generated_data_files})
...
+    add_custom_command(
+        OUTPUT
+            ${CMAKE_CURRENT_SOURCE_DIR}/src/test_keys.h
+        WORKING_DIRECTORY
+            ${CMAKE_CURRENT_SOURCE_DIR}
+        COMMAND
+            ${MBEDTLS_PYTHON_EXECUTABLE}
+                "${CMAKE_CURRENT_SOURCE_DIR}/scripts/generate_test_keys.py"
+                    "--output"
+                        "${CMAKE_CURRENT_SOURCE_DIR}/src/test_keys.h"
```
Comment (mpg) [inline tests/CMakeLists.txt:86]: Pre-existing and minor, but I find it weird to increase the indentation level for every line. 
```
        COMMAND
            ${MBEDTLS_PYTHON_EXECUTABLE}
            "${CMAKE_CURRENT_SOURCE_DIR}/scripts/generate_test_keys.py"
            "--output"
            "${CMAKE_CURRENT_SOURCE_DIR}/src/test_keys.h"
        DEPENDS
```
would make more sense to me. (Conceptually it's a flat list that's going to be passed to `exec()` or equivalent.)

File: tests/scripts/generate_test_keys.py [hunk 7]
```diff
@@ -12,6 +12,7 @@
 import argparse
 import scripts_path # pylint: disable=unused-import
 from mbedtls_dev.asymmetric_key_data import ASYMMETRIC_KEY_DATA
+from mbedtls_dev.build_tree import guess_project_root
 
 OUTPUT_HEADER_FILE = os.path.dirname(os.path.abspath(__file__)) + "/../src/test_keys.h"
```
Comment (mpg) [inline tests/scripts/generate_test_keys.py:17]: I think we should remove this line, as it's superseded by `default_output_path` in `main()` (and also it's not been updated to the "right" way of handling paths).

**Review by mpg — COMMENTED** [review]
Looking pretty good, thanks! Just a few things, mostly minor. The only one that really matters is if we should always re-generate or skip if the file is present.

File: tests/scripts/generate_test_keys.py [hunk 8]
```diff
@@ -99,9 +99,9 @@ def main() -> None:
     args = argparser.parse_args()
 
     output_file = args.output
-    # Remove output file if already existing.
+    # If the output file already exists, then we can quit (successfully)
```
[the remark this answers is not in the document]
Comment (valeriosetti) [reply tests/scripts/generate_test_keys.py:103]: Yes, I agree. I designed it that way though because I was afraid that regenerating all the times could cause some extra unnecessary work for the CI. But I agree that is the safest approach.

File: tests/scripts/generate_test_keys.py [hunk 9]
```diff
@@ -100,10 +100,6 @@ def main() -> None:
     args = argparser.parse_args()
 
     output_file = args.output
-    # If the output file already exists, then we can quit (successfully)
-    if os.path.exists(output_file):
-        return
-
     output_file = open(output_file, 'at')
```
Comment (mpg) [inline tests/scripts/generate_test_keys.py:102]: Pre-existing: this opens the file [in append mode](https://docs.python.org/3/library/functions.html#open). Seems to me we want `'wt'` instead in order to truncate it if it exists.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks almost good to me, except handling of the case where the file already exists - which the CI is also complaining about.

File: tests/scripts/generate_test_cert_macros.py [hunk 10]
```diff
@@ -14,51 +14,61 @@
...
 def main():
     parser = argparse.ArgumentParser()
-    build_argparser(parser)
+    default_output_path = os.path.join(THIS_DIR, '..', 'test_certs.h')
+    parser.add_argument('--output', type=str, default=default_output_path)
+    parser.add_argument('--list-dependencies', action='store_true')
     args = parser.parse_args()
-    return generate(**vars(args))
+
+    if args.list_dependencies is True:
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_test_cert_macros.py:60]:
```suggestion
    if args.list_dependencies
```

File: tests/scripts/generate_test_keys.py [hunk 11]
```diff
@@ -157,6 +162,7 @@ def main() -> None:
 """)
     output_file.write("\n".join(look_up_table))
     output_file.write("\n};\n")
+    output_file.flush()
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_test_keys.py:165]: Calling `flush` is only necessary if something is reading from the file while it's being generated. And that isn't happening. So why did you add it? This makes me worried that it's hiding some unrelated problem.

File: tests/scripts/generate_test_keys.py [hunk 12]
```diff
@@ -89,12 +90,16 @@ def get_look_up_table_entry(key_type: str, group_id_or_keybits: str,
     yield "      {0}, sizeof({0}),\n".format(priv_array_name)
     yield "      {0}, sizeof({0}) }},".format(pub_array_name)
 
+#pylint: disable=too-many-locals
 def main() -> None:
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_test_keys.py:94]: Please bypass pylint checks only at the narrowest practical level. Putting the directive at the toplevel applies to all subsequent functions as well.
```suggestion
def main() -> None:
    #pylint: disable=too-many-locals
```
Ideally we'd split this function: I think pylint's warning is justified here. But it was already just barely under the threshold before, so that is preexisting and out of scope.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Just a few minor things

File: tests/scripts/generate_test_keys.py (hunk 9 above)
Comment (valeriosetti) [reply tests/scripts/generate_test_keys.py:102]: Besides this was also causing the failure on "check_generated_files". Thanks for catching this :)

File: tests/scripts/generate_test_keys.py (hunk 11 above)
Comment (valeriosetti) [reply tests/scripts/generate_test_keys.py:165]: I had to add this `flush` [recently](https://github.com/Mbed-TLS/mbedtls/pull/9017/commits/0ddab0ecee4642e13706328588056aaa2121c540) because the CI was failing complaining about the fact that the final look-up table in `test_keys.h` was not defined. I tried also it locally and indeed it was failing for the same reason (but the script itself was not failing). Adding the `flush` solved the problem so I kept it.

Comment (gilles-peskine-arm) [reply tests/scripts/generate_test_keys.py:165]: Ok, so diving a bit deeper, the problem is that the script doesn't close the file explicitly. Closing the file does two things: it flushes the output buffer, and it throws an error if that last write can't be done (e.g. disk full).

I think flushing does a good enough job that the code is now correct. On the other hand, from what I've read ([1](https://realpython.com/why-close-file-python/#what-happens-if-you-dont-close-a-file-and-python-crashes) [2](https://stackoverflow.com/questions/17781906/what-happens-if-i-read-a-file-without-closing-it-afterwards)), flushing is necessary with some Python implementations, but not with CPython (where the file gets closed when `main` returns and `output_file` goes out of scope, and that throws an exception if the closure fails). And I think our CI only has CPython, so that wouldn't explain why adding `flush` was necessary to fix a practical problem. So I'm probably missing something, and I'm not satisfied with having `flush`.

Please use the [standard Python idiom](https://realpython.com/why-close-file-python/#toc) for working with a file:
```
with open(…) as output_file:
    output_file.write(…)
    …
```

(And if that indents the code too much, that's a sign that you should create an auxiliary function.)

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM except for the file operation scoping

**Comment by gilles-peskine-arm** [comment]
Please backport to 3.6. This file will move to the framework repository eventually, and in the meantime it's easier if we don't let the branches diverge.

**Comment by valeriosetti** [comment]
> Please backport to 3.6. This file will move to the framework repository eventually, and in the meantime it's easier if we don't let the branches diverge.

Sure, I was only waiting for this PR to get to a good point before starting it. Here it ís: https://github.com/Mbed-TLS/mbedtls/pull/9102

File: tests/scripts/generate_test_keys.py [hunk 13]
```diff
@@ -89,6 +89,32 @@ def get_look_up_table_entry(key_type: str, group_id_or_keybits: str,
     yield "      {0}, sizeof({0}),\n".format(priv_array_name)
     yield "      {0}, sizeof({0}) }},".format(pub_array_name)
 
+
+def write_output_file(output_file_name: str, arrays: List[str], look_up_table: List[str]):
+    with open(output_file_name, 'wt') as output:
+        output.write("""
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_test_keys.py:95]: Minor: there's now a blank line at the beginning of the file. You can avoid this with a backslash:
```suggestion
        output.write("""\
```

File: tests/scripts/generate_test_keys.py [hunk 14]
```diff
@@ -89,6 +89,32 @@ def get_look_up_table_entry(key_type: str, group_id_or_keybits: str,
...
+    const unsigned char *priv_key;
+    size_t priv_key_len;
+    const unsigned char *pub_key;
+    size_t pub_key_len;
+}};
+
+struct predefined_key_element predefined_keys[] = {{
+{}
+}};
+""".format("\n".join(look_up_table)))
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_test_keys.py:115]: Minor: it would be nice to have an `/* End of generated file */` comment. When debugging build scripts, it's sometimes useful to have confirmation that the file was fully generated. We currently have end-of-generated-file comments in some generated files, but not all.

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

File: tests/scripts/check-generated-files.sh [hunk 15]
```diff
@@ -132,6 +132,7 @@ check tests/scripts/generate_bignum_tests.py $(tests/scripts/generate_bignum_tes
 check tests/scripts/generate_ecp_tests.py $(tests/scripts/generate_ecp_tests.py --list)
 check tests/scripts/generate_psa_tests.py $(tests/scripts/generate_psa_tests.py --list)
 check tests/scripts/generate_test_keys.py tests/src/test_keys.h
+check tests/scripts/generate_test_cert_macros.py tests/src/test_certs.h
```
Comment (mpg) [inline tests/scripts/check-generated-files.sh:135]: I don't think this is the right place: we're in the list of things common to Mbed TLS and TF-PSA-Crypto, but certs (X.509) are part of Mbed TLS, not PSA Crypto, so this line should probably go in "Additional checks for Mbed TLS only" below.

File: tests/scripts/generate_test_keys.py [hunk 16]
```diff
@@ -89,6 +89,32 @@ def get_look_up_table_entry(key_type: str, group_id_or_keybits: str,
...
+
+
 def main() -> None:
     #pylint: disable=too-many-locals
```
Comment (mpg) [inline tests/scripts/generate_test_keys.py:121]: Question: can we remove this line now that writing to the output file has been moved to a separate function? (No problem if we can't, just checking.)

**Review by mpg — CHANGES_REQUESTED** [review]
Almost there, just one thing in `check-generated-files.sh`.

File: tests/scripts/generate_test_keys.py (hunk 16 above)
Comment (valeriosetti) [reply tests/scripts/generate_test_keys.py]: Ops sorry, I was sure I added this change in https://github.com/Mbed-TLS/mbedtls/pull/9017/commits/2c5899664e0d38076f18471bcda1219c5141a800 (even that commit's message reports this), but apparently I missed that somewhere. Anyway I updated that commit (new version is https://github.com/Mbed-TLS/mbedtls/pull/9017/commits/cc403cb6ec2a8ad4aa8ef02efa753d1d643e0df4) so this should be fine now

**Review by mpg — DISMISSED** [review]
LGTM, thanks!

File: tests/scripts/check-generated-files.sh [hunk 17]
```diff
@@ -139,6 +139,8 @@ if in_mbedtls_repo; then
     check scripts/generate_query_config.pl programs/test/query_config.c
     check scripts/generate_features.pl library/version_features.c
     check scripts/generate_ssl_debug_helpers.py library/ssl_debug_helpers_generated.c
+    check tests/scripts/generate_test_keys.py tests/src/test_keys.h
```
Comment (mpg) [inline tests/scripts/check-generated-files.sh:142]: Note: I'm not super sure about this line (my previous comment was only about `test_certs.h`, so I'm asking on slack.

Comment (gilles-peskine-arm) [reply tests/scripts/check-generated-files.sh:142]: `test_keys.h` is used by crypto (`pk.h`, likely some day by a PSA equivalent API), so it belongs in the crypto+TLS section. But that's not really important: the sections in this file date back from before we had planned the framework repository, so they'll need to be revised anyway.

Comment (gilles-peskine-arm) [reply tests/scripts/check-generated-files.sh:142]: So it's not critical. But it would help to get the section right, because the next person to work on this file might be dealing with lost of unfamiliar files, so we should try to give them clues as to what to put where.

**Review by mpg — APPROVED** [review]
LGTM

File: tests/scripts/generate_test_keys.py [hunk 18]
```diff
@@ -89,25 +89,51 @@ def get_look_up_table_entry(key_type: str, group_id_or_keybits: str,
...
+
+struct predefined_key_element predefined_keys[] = {{
+{}
+}};
+
+/* End of generated file */
+""".format("\n".join(look_up_table)))
+
+
 def main() -> None:
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_test_keys.py:120]: There are still too many locals according to pylint, and I tend to agree that this function would read better with another auxiliary function like `arrays, look_up_table = collect_keys()`

File: tests/scripts/generate_test_keys.py [hunk 19]
```diff
@@ -116,16 +116,10 @@ def write_output_file(output_file_name: str, arrays: List[str], look_up_table: L
...
-
-    argparser = argparse.ArgumentParser()
-    argparser.add_argument("--output", help="Output file", default=default_output_path)
-    args = argparser.parse_args()
-
-    output_file = args.output
-
+def collect_keys() -> Tuple[List[str], List[str]]:
+    """"
+    asdf
```
[the remark this answers is not in the document]
Comment (valeriosetti) [reply tests/scripts/generate_test_keys.py]: Ops, sorry, that was only meant to test `check_python_files` of course, but then I forgot to update it before pushing the commit :/

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by mpg — APPROVED** [review]
LGTM

**Comment by mpg** [comment]
I think we should have added the now-generated files to `.gitignore` as well...
